From 07965240cdf318174041195d7a48a1102efe3847 Mon Sep 17 00:00:00 2001
From: Yang Yingliang <yangyingliang@huawei.com>
Date: Tue, 2 Mar 2021 17:16:28 +0800
Subject: [PATCH] cgroup: disable kernel memory accounting for all memory
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 07965240cdf318174041195d7a48a1102efe3847
Modified-by-SEL: Yes, refreshed due to context change

 cgroups by default

hulk inclusion
category: bugfix
Bugzilla: 50424
CVE: NA

----------------------------------------

The kernel memory accounting for all memory cgroups is
not stable, and it will cause a 100% regression in
hackbench compared with kernel-4.19, so disable it by
default. We can use the following command line to enable
or disable it:
cgroup.memory=kmem or cgroup.memory=kmem.

Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Reviewed-by: Kefeng Wang <wangkefeng.wang@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
Signed-off-by: Kai Liu <kai.liu@suse.com>
---
 Documentation/admin-guide/cgroup-v1/memory.rst  |    6 +++---
 Documentation/admin-guide/kernel-parameters.txt |    1 +
 mm/memcontrol.c                                 |    6 ++++--
 3 files changed, 8 insertions(+), 5 deletions(-)

--- a/Documentation/admin-guide/cgroup-v1/memory.rst
+++ b/Documentation/admin-guide/cgroup-v1/memory.rst
@@ -305,9 +305,9 @@ the amount of kernel memory used by the
 different than user memory, since it can't be swapped out, which makes it
 possible to DoS the system by consuming too much of this precious resource.
 
-Kernel memory accounting is enabled for all memory cgroups by default. But
-it can be disabled system-wide by passing cgroup.memory=nokmem to the kernel
-at boot time. In this case, kernel memory will not be accounted at all.
+Kernel memory accounting is disabled for all memory cgroups by default. But
+it can be enabled system-wide by passing cgroup.memory=kmem to the kernel
+at boot time. In this case, kernel memory will all be accounted.
 
 Kernel memory limits are not imposed for the root cgroup. Usage for the root
 cgroup may or may not be accounted. The memory used is accumulated into
--- a/Documentation/admin-guide/kernel-parameters.txt
+++ b/Documentation/admin-guide/kernel-parameters.txt
@@ -517,6 +517,7 @@
 			Format: <string>
 			nosocket -- Disable socket memory accounting.
 			nokmem -- Disable kernel memory accounting.
+			kmem -- Enable kernel memory accounting.
 
 	checkreqprot	[SELINUX] Set initial checkreqprot flag value.
 			Format: { "0" | "1" }
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -82,8 +82,8 @@ DEFINE_PER_CPU(struct mem_cgroup *, int_
 /* Socket memory accounting disabled? */
 static bool cgroup_memory_nosocket;
 
-/* Kernel memory accounting disabled? */
-static bool cgroup_memory_nokmem;
+/* Kernel memory accounting disabled */
+static bool cgroup_memory_nokmem = true;
 
 /* Whether the swap controller is active */
 #ifdef CONFIG_MEMCG_SWAP
@@ -7084,6 +7084,8 @@ static int __init cgroup_memory(char *s)
 			cgroup_memory_nosocket = true;
 		if (!strcmp(token, "nokmem"))
 			cgroup_memory_nokmem = true;
+		else if (!strcmp(token, "kmem"))
+			cgroup_memory_nokmem = false;
 	}
 	return 1;
 }
